src/shogun/evaluation/Calibration.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 2011-2012 Heiko Strathmann
+ * Copyright (C) 2011 Berlin Institute of Technology and Max-Planck-Society
+ */
+
+#include <shogun/evaluation/Calibration.h>
+#include <shogun/evaluation/CalibrationMethod.h>
+#include <shogun/lib/config.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/mathematics/Statistics.h>
+
+using namespace shogun;
+
+CBinaryLabels* CCalibration::apply_binary(CFeatures* features)
+{
+
+	CBinaryLabels* result_labels = (CBinaryLabels*)m_machine->apply(features);
+	CCalibrationMethod* method =
+	    (CCalibrationMethod*)m_calibration_machines->get_element(0);
+	SGVector<float64_t> confidence_values =
+	    method->apply_binary(result_labels->get_values());
+	result_labels->set_values(confidence_values);
+
+	return result_labels;
+}
+
+CMulticlassLabels* CCalibration::apply_multiclass(CFeatures* features)
+{
+	index_t num_calibration_machines =
+	    m_calibration_machines->get_num_elements();
+	CMulticlassLabels* result_labels =
+	    (CMulticlassLabels*)m_machine->apply(features);
+	for (index_t i = 0; i < num_calibration_machines; ++i)
+	{
+		CCalibrationMethod* method =
+		    (CCalibrationMethod*)m_calibration_machines->get_element(i);
+		SGVector<float64_t> confidence_values =
+		    method->apply_binary(result_labels->get_multiclass_confidences(i));
+		result_labels->set_multiclass_confidences(i, confidence_values);
+	}
+
+	SGVector<float64_t> temp_confidences;
+	index_t num_classes = result_labels->get_num_classes();
+	index_t num_samples;
+
+	// normalize the probabilities
+	for (index_t i = 0; i < num_classes; ++i)
+	{
+		SGVector<float64_t> confidence_values =
+		    result_labels->get_multiclass_confidences(i);
+		if (i == 0)
+		{
+			temp_confidences = confidence_values;
+			num_samples = temp_confidences.vlen;
+		}
+		else
+		{
+			for (index_t j = 0; j < num_samples; ++j)
+			{
+				temp_confidences[j] += confidence_values[j];
+			}
+		}
+	}
+	for (index_t i = 0; i < num_classes; ++i)
+	{
+		SGVector<float64_t> confidence_values =
+		    result_labels->get_multiclass_confidences(i);
+		for (index_t j = 0; j < num_samples; ++j)
+		{
+			confidence_values[j] = confidence_values[j] / temp_confidences[j];
+		}
+		result_labels->set_multiclass_confidences(i, confidence_values);
+	}
+
+	return result_labels;
+}
+
+CMulticlassLabels*
+CCalibration::apply_locked_multiclass(SGVector<index_t> subset_indices)
```

Comment:
this method is almost the copy-paste of `CCalibration::apply_multiclass`. you can basically add the same parts to a common function and use that

Reply:
I am also confused, why is this necessary?

Reply:
I have now used a template function to hold the common code. I'll make a pr shortly